File: 09-backtracking/00051-N_Queens/00051-n_queens.go
// 51: N Queens
// https://leetcode.com/problems/n-queens

package main

import (
	"fmt"
	"strings"
)

func backtrack(n int, row int, nQueens []string, result *[][]string) {
	if row == n {
		t := make([]string, len(nQueens))
		copy(t, nQueens)
		*result = append(*result, t)
		return
	}

	for col := 0; col < n; col++ {
		if isValid(nQueens, row, col, n) {
			r := []rune(nQueens[row])
			r[col] = 'Q'
			nQueens[row] = string(r)
			backtrack(n, row+1, nQueens, result)
			r = []rune(nQueens[row])
			r[col] = '.'
			nQueens[row] = string(r)
		}
	}
}

func isValid(nQueens []string, row, col, n int) bool {
	for i := 0; i < row; i++ {
		if nQueens[i][col] == 'Q' {
			return false
		}
	}

	for i, j := row-1, col-1; i >= 0 && j >= 0; i, j = i-1, j-1 {
		if nQueens[i][j] == 'Q' {
			return false
		}
	}

	for i, j := row-1, col+1; i >= 0 && j < n; i, j = i-1, j+1 {
		if nQueens[i][j] == 'Q' {
			return false
		}
	}

	return true
}

// SOLUTION
func solveNQueens(n int) [][]string {
	nQueens := make([]string, n)
	for i := 0; i < n; i++ {
		nQueens[i] = strings.Repeat(".", n)
	}
	var result [][]string
	backtrack(n, 0, nQueens, &result)
	return result
}

func main() {
	// INPUT
	n := 4

	// OUTPUT
	result := solveNQueens(n)

	fmt.Println(result)
}

File: 09-backtracking/00051-N_Queens/00051-n_queens.py
# 51: N Queens
# https://leetcode.com/problems/n-queens


from typing import List

class Solution:
    # SOLUTION
    def solveNQueens(self, n: int) -> List[List[str]]:
        result = []
        nQueens = [("." * n)] * n

        def isValid(row: int, col: int) -> bool:
            for i in range(row):
                if nQueens[i][col] == 'Q':
                    return False

            i, j = row-1, col-1
            while i>=0 and j>=0:
                if nQueens[i][j] == 'Q':
                    return False
                i -= 1
                j -= 1

            i, j = row-1, col+1
            while i>=0 and j<n:
                if nQueens[i][j] == 'Q':
                    return False
                i -= 1
                j += 1

            return True


        def backtrack(row: int) -> None:
            if (row == n):
                result.append(nQueens.copy())
                return

            for col in range(n):
                if isValid(row, col):
                    r =  nQueens[row]
                    r = r[0:col] + "Q" + r[col+1:]
                    nQueens[row] = r
                    backtrack(row+1)
                    r = r[0:col] + "." + r[col+1:]
                    nQueens[row] = r

        backtrack(0)
        return result



if __name__ == "__main__":
    o = Solution()

    # INPUT
    n = 4

    # OUTPUT
    result = o.solveNQueens(n)

    print(result)
